core/node/auth/auth_impl.go

```diff
@@ -1250,9 +1264,102 @@ func (ca *chainAuth) checkIsApp(
 		}
 	}
 
+	if args.kind == chainAuthKindDmValidation {
+		return ca.checkDmValidation(ctx, cfg, args)
+	}
+
 	return nil, nil
 }
 
+func (ca *chainAuth) checkDmValidation(
+	ctx context.Context,
+	cfg *config.Config,
+	args *ChainAuthArgs,
+) (CacheResult, error) {
+	log := logging.FromCtx(ctx)
+
+	firstPartyIsApp, firstPartyAppContract, err := ca.appRegistryContract.UserIsRegisteredAsApp(ctx, args.principal)
+	if err != nil {
+		return nil, err
+	}
+
+	secondPartyIsApp, secondPartyAppContract, err := ca.appRegistryContract.UserIsRegisteredAsApp(
+		ctx,
+		args.botClientAddress,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// User-to-user
+	if !firstPartyIsApp && !secondPartyIsApp {
+		if args.requireBotParty {
+			log.Debugw(
+				"checkDmValidation: bot required but neither party is a bot",
+				"firstParty", args.principal,
+				"secondParty", args.botClientAddress,
+			)
+			return boolCacheResult{false, EntitlementResultReason_NO_BOT_PARTY}, nil
+		}
+		log.Debugw(
+			"checkDmValidation: user-to-user DM",
+			"firstParty", args.principal,
+			"secondParty", args.botClientAddress,
+		)
+		return boolCacheResult{true, EntitlementResultReason_NONE}, nil
+	}
+
+	// Bot involved: check if the user has the bot installed
+	if firstPartyIsApp {
+		return ca.checkAppInstalledOnUser(ctx, cfg, args.botClientAddress, firstPartyAppContract)
+	}
+	return ca.checkAppInstalledOnUser(ctx, cfg, args.principal, secondPartyAppContract)
+}
+
+func (ca *chainAuth) checkAppInstalledOnUser(
+	ctx context.Context,
+	cfg *config.Config,
+	userAddress common.Address,
+	appContractAddress common.Address,
+) (CacheResult, error) {
+	log := logging.FromCtx(ctx)
+
+	userArgs := &ChainAuthArgs{principal: userAddress}
+	wallets, err := ca.getLinkedWallets(ctx, cfg, userArgs)
```

Comment:
I think we already called get linked wallets, but i can't parse this code any more. hopefully this will return a cached version